File: 105.cpp
#include <unordered_map>
struct Solution {
    TreeNode *gen(
            int pi, int pj, int ii, int ij, 
            vector<int>& preorder, vector<int>& inorder,
            unordered_map<int, int> &h){
        if (pi >= pj){
            return nullptr;
        }
        int val = preorder[pi];
        int l = h[val] - ii;
        return new TreeNode(
                val,
                this->gen(pi + 1, pi + l + 1, ii, ii + l, preorder, inorder, h),
                this->gen(pi + l + 1, pj, ii + l + 1, ij, preorder, inorder, h)
        );
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = preorder.size();
        unordered_map<int, int> h;
        for (int i = 0; i < n; ++i){
            h[inorder[i]] = i;
        }
        return this->gen(0, n, 0, n, preorder, inorder, h);
    }
};

File: sol/105/105.py
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        h = dict()
        for i, x in enumerate(inorder):
            h[x] = i
        n = len(preorder)
        def gen(pi, pj, ii, ij):
            if pi >= pj:
                return None
            val = preorder[pi]
            l = h[val] - ii
            return TreeNode(
                    val,
                    gen(pi + 1, pi + l + 1, ii, ii + l),
                    gen(pi + l + 1, pj, ii + l + 1, ij)
            )
        return gen(0, n, 0, n)
